File: app/src/main/java/com/example/services/AgingService.kt
package com.example.services

import com.example.data.model.ClientCharacter
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * AGING SERVICE (Prompt 02 of 20: CHARACTER SYSTEM)
 *
 * Implements biological aging and progressive degradation:
 * - Health (gradual degradation, scaling in senior years)
 * - Fertility (steady in youth, steep drop in late 30s/40s)
 * - Energy (gradual systemic decline)
 * - Athletic Performance (mild until 30, SHARP drop past age 30)
 *
 * ARCHITECTURAL MANDATE:
 * Must route degradation adjustments through the central FormulaEngine rather
 * than executing inline math.
 */
data class AgingResult(
    val newAge: Int,
    val previousHealth: Int,
    val previousFertility: Int,
    val previousEnergy: Int,
    val previousAthleticPerformance: Int,
    val newHealth: Int,
    val newFertility: Int,
    val newEnergy: Int,
    val newAthleticPerformance: Int,
    val narrative: String
)

object AgingService {

    fun ageOneYear(
        character: ClientCharacter,
        currentAge: Int,
        worldStability: Double = 1.0,
        economicCycle: Double = 1.0,
        luckRoll: Double? = null
    ): AgingResult {
        val nextAge = currentAge + 1

        // 1. Evaluate biological resilience via central FormulaEngine
        val foundation = FormulaEngine.CharacterFoundation(
            intelligence = character.intelligence.toDouble(),
            discipline = character.discipline.toDouble(),
            willpower = character.willpower.toDouble(),
            ambition = character.ambition.toDouble(),
            health = character.health.toDouble(),
            looks = character.looks.toDouble(),
            smarts = character.smarts.toDouble(),
            happiness = character.happiness.toDouble()
        )

        val momentum = FormulaEngine.Momentum(
            trajectoryScore = if (character.health > 70) 0.2 else -0.2,
            fatigueFactor = max(0.0, (100 - character.energy) / 100.0)
        )

        val world = FormulaEngine.WorldVariables(
            economicCycleIndex = economicCycle,
            stabilityIndex = worldStability
        )

        val luck = luckRoll?.let { FormulaEngine.Luck(rawRoll = it) } ?: FormulaEngine.Luck()

        val formulaOutcome = FormulaEngine.evaluate(
            characterFoundation = foundation,
            momentum = momentum,
            worldVariables = world,
            luck = luck
        )

        // Composite resilience factor: 0.65 (high willpower/discipline/health) to 1.35
        val resilienceScore = formulaOutcome.compositeScore
        val resilienceFactor = (1.45 - (resilienceScore / 100.0)).coerceIn(0.65, 1.35)

        // 2. Health degradation
        val baseHealthLoss = when {
            nextAge > 80 -> 4.5
            nextAge > 65 -> 3.0
            nextAge > 45 -> 1.6
            nextAge > 25 -> 0.7
            else -> 0.1
        }
        val healthLoss = (baseHealthLoss * resilienceFactor).roundToInt()
        val newHealth = (character.health - healthLoss).coerceIn(0, 100)

        // 3. Fertility degradation
        val baseFertilityLoss = when {
            nextAge > 45 -> 8.0
            nextAge > 36 -> 4.0
            nextAge > 28 -> 1.5
            else -> 0.2
        }
        val fertilityLoss = (baseFertilityLoss * resilienceFactor).roundToInt()
        val newFertility = (character.fertility - fertilityLoss).coerceIn(0, 100)

        // 4. Energy degradation
        val baseEnergyLoss = when {
            nextAge > 70 -> 3.5
            nextAge > 50 -> 2.0
            nextAge > 30 -> 1.0
            else -> 0.4
        }
        val energyLoss = (baseEnergyLoss * resilienceFactor).roundToInt()
        val newEnergy = (character.energy - energyLoss).coerceIn(0, 100)

        // 5. Athletic Performance degradation (Prompt 02: SHARP DROP past age 30)
        val baseAthleticLoss = when {
            nextAge > 55 -> 8.0
            nextAge > 40 -> 6.0
            nextAge > 30 -> 4.5 // Sharp degradation past 30
            nextAge > 25 -> 1.2
            else -> 0.2
        }
        val athleticLoss = (baseAthleticLoss * resilienceFactor).roundToInt()
        val newAthletic = (character.athleticPerformance - athleticLoss).coerceIn(0, 100)

        val narrativeBuilder = StringBuilder("Reached age $nextAge.")
        if (nextAge > 30 && athleticLoss >= 4) {
            narrativeBuilder.append(" Athletic speed and stamina experienced a sharp post-30 decline (-$athleticLoss ATH).")
        }
        if (healthLoss > 2) {
            narrativeBuilder.append(" Natural aging reduced baseline vitality (-$healthLoss HLT).")
        }
        if (resilienceScore >= 75.0) {
            narrativeBuilder.append(" Strong foundation and discipline cushioned physical degradation.")
        }

        return AgingResult(
            newAge = nextAge,
            previousHealth = character.health,
            previousFertility = character.fertility,
            previousEnergy = character.energy,
            previousAthleticPerformance = character.athleticPerformance,
            newHealth = newHealth,
            newFertility = newFertility,
            newEnergy = newEnergy,
            newAthleticPerformance = newAthletic,
            narrative = narrativeBuilder.toString()
        )
    }
}

File: lib/agingService.ts
/**
 * AGING SERVICE (Prompt 02 of 20: CHARACTER SYSTEM)
 *
 * Implements biological aging and progressive degradation for:
 * - Health (gradual degradation, scaling after middle age)
 * - Fertility (gradual in 20s/30s, steep decline in late 30s/40s)
 * - Energy (gradual systemic decline)
 * - Athletic Performance (mild until 30, SHARP drop past age 30)
 *
 * ARCHITECTURAL MANDATE:
 * Must call the central FormulaEngine rather than calculating degradation inline.
 * The FormulaEngine's composite evaluation (incorporating willpower, discipline,
 * foundation health, and world variables) provides stochastic biological resilience.
 */

import { evaluateFormula } from './formulaEngine';
import { AgingResult, ClientCharacter } from '../types';

export interface AgingOptions {
  worldStability?: number;
  economicCycle?: number;
  luckRoll?: number;
}

export class AgingService {
  /**
   * Evaluates one year of aging on a character.
   * Returns new stats, stat deltas, and narrative logging.
   */
  public static ageOneYear(
    character: Pick<
      ClientCharacter,
      | 'intelligence'
      | 'discipline'
      | 'willpower'
      | 'ambition'
      | 'health'
      | 'looks'
      | 'smarts'
      | 'happiness'
      | 'fertility'
      | 'energy'
      | 'athleticPerformance'
    >,
    currentAge: number,
    options: AgingOptions = {}
  ): AgingResult {
    const nextAge = currentAge + 1;

    // 1. Evaluate biological resilience via central FormulaEngine
    const formulaOutcome = evaluateFormula({
      characterFoundation: {
        intelligence: character.intelligence,
        discipline: character.discipline,
        willpower: character.willpower,
        ambition: character.ambition,
        health: character.health,
        looks: character.looks,
        smarts: character.smarts,
        happiness: character.happiness,
      },
      momentum: {
        trajectoryScore: character.health > 70 ? 0.2 : -0.2,
        fatigueFactor: Math.max(0, (100 - character.energy) / 100),
      },
      worldVariables: {
        stabilityIndex: options.worldStability ?? 1.0,
        economicCycleIndex: options.economicCycle ?? 1.0,
      },
      luck: {
        rawRoll: options.luckRoll,
      },
    });

    // High composite score mitigates degradation (resilience factor: 0.65 to 1.35)
    // Characters with high discipline, willpower, and foundation health preserve their stats better
    const resilienceScore = formulaOutcome.compositeScore;
    const resilienceFactor = Math.max(0.65, Math.min(1.35, 1.45 - (resilienceScore / 100)));

    // 2. Compute Health degradation
    let baseHealthLoss = 0;
    if (nextAge > 80) {
      baseHealthLoss = 4.5;
    } else if (nextAge > 65) {
      baseHealthLoss = 3.0;
    } else if (nextAge > 45) {
      baseHealthLoss = 1.6;
    } else if (nextAge > 25) {
      baseHealthLoss = 0.7;
    } else {
      baseHealthLoss = 0.1;
    }
    const healthLoss = Math.round(baseHealthLoss * resilienceFactor);
    const newHealth = Math.max(0, Math.min(100, character.health - healthLoss));

    // 3. Compute Fertility degradation
    let baseFertilityLoss = 0;
    if (nextAge > 45) {
      baseFertilityLoss = 8.0;
    } else if (nextAge > 36) {
      baseFertilityLoss = 4.0;
    } else if (nextAge > 28) {
      baseFertilityLoss = 1.5;
    } else {
      baseFertilityLoss = 0.2;
    }
    const fertilityLoss = Math.round(baseFertilityLoss * resilienceFactor);
    const newFertility = Math.max(0, Math.min(100, character.fertility - fertilityLoss));

    // 4. Compute Energy degradation
    let baseEnergyLoss = 0;
    if (nextAge > 70) {
      baseEnergyLoss = 3.5;
    } else if (nextAge > 50) {
      baseEnergyLoss = 2.0;
    } else if (nextAge > 30) {
      baseEnergyLoss = 1.0;
    } else {
      baseEnergyLoss = 0.4;
    }
    const energyLoss = Math.round(baseEnergyLoss * resilienceFactor);
    const newEnergy = Math.max(0, Math.min(100, character.energy - energyLoss));

    // 5. Compute Athletic Performance degradation (SHARP DROP past age 30)
    let baseAthleticLoss = 0;
    if (nextAge > 55) {
      baseAthleticLoss = 8.0;
    } else if (nextAge > 40) {
      baseAthleticLoss = 6.0;
    } else if (nextAge > 30) {
      // Prompt 02 Mandate: Sharp drop past age 30
      baseAthleticLoss = 4.5;
    } else if (nextAge > 25) {
      baseAthleticLoss = 1.2;
    } else {
      baseAthleticLoss = 0.2;
    }
    const athleticLoss = Math.round(baseAthleticLoss * resilienceFactor);
    const newAthletic = Math.max(0, Math.min(100, character.athleticPerformance - athleticLoss));

    // Narrative generation
    let narrative = `Turned age ${nextAge}.`;
    if (nextAge > 30 && athleticLoss >= 4) {
      narrative += ` Physical agility and athletic endurance took a noticeable post-30 dip (-${athleticLoss} ATH).`;
    }
    if (healthLoss > 2) {
      narrative += ` Aging toll reduced vitality (-${healthLoss} HLT).`;
    }
    if (resilienceScore >= 75) {
      narrative += ` High intrinsic discipline and resilience cushioned biological decline.`;
    }

    return {
      newAge: nextAge,
      previousStats: {
        health: character.health,
        fertility: character.fertility,
        energy: character.energy,
        athleticPerformance: character.athleticPerformance,
      },
      newStats: {
        health: newHealth,
        fertility: newFertility,
        energy: newEnergy,
        athleticPerformance: newAthletic,
      },
      deltas: {
        health: newHealth - character.health,
        fertility: newFertility - character.fertility,
        energy: newEnergy - character.energy,
        athleticPerformance: newAthletic - character.athleticPerformance,
      },
      narrative,
    };
  }
}
